examples/econais/ec19d/sdk_example_kaa_client/src/kaa_client.c

```diff
@@ -0,0 +1,823 @@
+                                             , KAA_SERVICE_LOGGING};
+static const int OPERATIONS_SERVICES_COUNT = sizeof(OPERATIONS_SERVICES) / sizeof(kaa_service_t);
+
+void thread_run_fn(uintptr_t arg);
+
+kaa_error_t kaa_client_create(kaa_client_t **kaa_client, kaa_client_props_t *props)
+{
+    KAA_RETURN_IF_NIL2(kaa_client, props, KAA_ERR_BADPARAM);
+
+    kaa_error_t error_code = KAA_ERR_NONE;
+
+    error_code = kaa_init_security_stuff();
+    KAA_RETURN_IF_ERR(error_code);
+
+    kaa_client_t *self = sndc_mem_calloc(1,sizeof(kaa_client_t));
+    KAA_RETURN_IF_NIL(self,KAA_ERR_NOMEM);
+
+    self->thread_name = sndc_mem_strdup("Kaa-Client-Thread");
+    self->operate = true;
+    self->max_update_time = props->max_update_time;
+
+    print_mem_stat(self);
+    sndc_printf("Initializing Kaa SDK...\n");
+    sndc_thrd_delay(TRACE_DELAY * SNDC_MILLISECOND);
+
+    error_code = kaa_init(&self->kaa_context);
+    if (error_code) {
+        sndc_printf("Error during Kaa context creation %d\n", error_code);
+        sndc_thrd_delay(TRACE_DELAY * SNDC_MILLISECOND);
+        goto error;
+    }
+
+    KAA_LOG_INFO(self->kaa_context->logger, KAA_ERR_NONE, "Kaa framework initialized.");
+
+    print_mem_stat(self);
+    error_code = kaa_log_collector_init(self);
+    if (error_code) {
+        sndc_printf("Failed to init Kaa log collector %d\n", error_code);
+        sndc_thrd_delay(TRACE_DELAY * SNDC_MILLISECOND);
+        goto error;
+    }
+
+    KAA_LOG_INFO(self->kaa_context->logger, KAA_ERR_NONE, "Kaa log collector initialized.");
+
+    *kaa_client = self;
+
+    /* initialize semaphore */
+    sndc_sem_init(&self->start_semophore, 0);
+    sndc_sem_init(&self->logging_semophore, 0);
+
+    int status = sndc_thrd_create(
+            self->thread_name,
+            thread_run_fn,
+            (uintptr_t)self,
+            SNDC_THRD_PRIORITY_DEFAULT,
+            THREAD_STACKSIZE_DMR_START); //Defined 4K-8 bytes of stack
+
+    switch (status) {
+        case 0:
+
+            break;
+        case -EINVAL:
+            error_code = KAA_ERR_BADPARAM;
+            break;
+        case -ENOMEM:
+            error_code = KAA_ERR_NOMEM;
+            break;
+        default:
+            error_code = KAA_ERR_BADDATA;
+            break;
+    }
+
+    return error_code;
+
+error:
+
+    sndc_printf("Kaa initialization failed. error_code %d\n", error_code);
+    sndc_thrd_delay(TRACE_DELAY * SNDC_MILLISECOND);
+    kaa_client_destroy(self);
+
+    return error_code;
+}
+
+void kaa_client_destroy(kaa_client_t *self)
+{
+    if(!self)
+        return;
+
+    if (self->start_semophore) {
+        sndc_sem_destroy(self->start_semophore);
+        self->start_semophore = NULL;
+    }
+
+    if (self->logging_semophore) {
+        sndc_sem_destroy(self->logging_semophore);
+        self->logging_semophore = NULL;
+    }
+
+    if (self->operations_channel.context) {
+        self->operations_channel.destroy(self->operations_channel.context);
```

Comment:
Is it possible that channel may be de-destroyed twice as the SDK' channel manager does the same?